sqld/src/test/bottomless.rs

```diff
@@ -374,49 +381,39 @@ impl Drop for DbFileCleaner {
     }
 }
 
-/// Guardian struct used for cleaning up the test data from
-/// S3 bucket dir at the beginning and end of a test.
-struct S3BucketCleaner(&'static str);
-
-impl S3BucketCleaner {
-    async fn new(bucket: &'static str) -> Self {
-        let _ = Self::cleanup(bucket).await; // cleanup the bucket before test
-        S3BucketCleaner(bucket)
-    }
+/// Clean up the test data from S3 bucket dir at the beginning and end of a test.
+async fn with_s3_bucket_cleanup<T>(bucket: &'static str, fut: impl Future<Output = T>) -> T {
```

Comment:
My main but hurt here is that it won't work when panics happen (which is frequent in tests due to assertions).

Reply:
You're right, I'll restore the guard and try something else :)

Reply:
@Horusiath I found two solutions using the guard:

The first one is to create a runtime on another thread. The reason `futures::executor::block_on` hanged was because it blocks the thread that tokio uses to drive their tasks.
```rust
impl Drop for S3BucketCleaner {
     fn drop(&mut self) {
         std::thread::spawn(|| {
             let _ = tokio::runtime::Builder::new_current_thread()
                 .enable_all()
                 .build()
                 .unwrap()
                 .block_on(Self::cleanup(self.0));
         })
         .join()
         .unwrap();
     }
}
```

The other solution is to use [`tokio::task::block_in_place`](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html) with a multi-thread scheduler:
```rust
impl Drop for S3BucketCleaner {
    fn drop(&mut self) {
        tokio::task::block_in_place(|| {
            let _  = tokio::runtime::Handle::current().block_on(Self::cleanup(self.0));
        });
    }
}
```
But it requires the tests to be run with:
```rust
#[tokio::test(flavor = "multi_thread")]
```

If any of those solutions satisfies you, I can update my PR 😄.

Thanks!

Reply:
@Horusiath I found two solutions using the guard:

The first one is to create a runtime on another thread. The reason futures::executor::block_on hanged was because it blocks the thread that tokio uses to drive their tasks.

```rust
impl Drop for S3BucketCleaner {
     fn drop(&mut self) {
         std::thread::spawn(|| {
             let _ = tokio::runtime::Builder::new_current_thread()
                 .enable_all()
                 .build()
                 .unwrap()
                 .block_on(Self::cleanup(self.0));
         })
         .join()
         .unwrap();
     }
}
```

The other solution is to use tokio::task::block_in_place with a multi-thread scheduler:
```rust
impl Drop for S3BucketCleaner {
    fn drop(&mut self) {
        tokio::task::block_in_place(|| {
            let _  = tokio::runtime::Handle::current().block_on(Self::cleanup(self.0));
        });
    }
}
```

But it requires the tests to be run with:
```rust
#[tokio::test(flavor = "multi_thread")]
```
If any of those solutions satisfies you, I can update my PR 😄.

Thanks!

Reply:
I'd say that multi_thread scheduler might perform better given the fair amount to extra fibres created by sqld server and running in these tests.

Reply:
Done!

Reply:
Done!